org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/GradleClasspathContainerRuntimeClasspathEntryResolver.java

```diff
@@ -84,17 +110,62 @@ private void collectContainerRuntimeClasspath(IClasspathContainer container, Lis
+                        collectContainerRuntimeClasspathWithGradleScopes(dependencyProject, result, true, configurationScopes);
                     }
                 }
             }
         }
     }
 
+    private IRuntimeClasspathEntry[] runtimeClasspathWithTestSources(IJavaProject project, boolean excludeTestCode) throws CoreException {
+        List<IRuntimeClasspathEntry> result = Lists.newArrayList();
+
+        IClasspathContainer container = JavaCore.getClasspathContainer(GradleClasspathContainer.CONTAINER_PATH, project);
+        if (container == null) {
+            return new IRuntimeClasspathEntry[0];
+        }
+
+        for (final IClasspathEntry cpe : container.getClasspathEntries()) {
+            if (cpe.getEntryKind() == IClasspathEntry.CPE_LIBRARY && !(excludeTestCode && hasTestAttribute(cpe))) {
+                result.add(JavaRuntime.newArchiveRuntimeClasspathEntry(cpe.getPath()));
+            } else if (cpe.getEntryKind() == IClasspathEntry.CPE_PROJECT) {
+                Optional<IProject> candidate = findAccessibleJavaProject(cpe.getPath().segment(0));
+                if (candidate.isPresent()) {
+                    IJavaProject dependencyProject = JavaCore.create(candidate.get());
+                    IRuntimeClasspathEntry projectRuntimeEntry = JavaRuntime.newProjectRuntimeClasspathEntry(dependencyProject);
+                    // add the project entry itself so that the source lookup can find the classes
+                    // see https://github.com/eclipse/buildship/issues/383
+                    result.add(projectRuntimeEntry);
+                    Collections.addAll(result, invokeJavaRuntimeResolveRuntimeClasspathEntry(projectRuntimeEntry, dependencyProject, excludeTestCode));
+                }
+            }
+        }
+        return result.toArray(new IRuntimeClasspathEntry[result.size()]);
+    }
+
+    private static IRuntimeClasspathEntry[] invokeJavaRuntimeResolveRuntimeClasspathEntry(IRuntimeClasspathEntry projectRuntimeEntry, IJavaProject dependencyProject, boolean excludeTestCode) throws CoreException{
+        // JavaRuntime.resolveRuntimeClasspathEntry is available since Eclipse 4.8
+        try {
+            Method method = JavaRuntime.class.getMethod("resolveRuntimeClasspathEntry", IRuntimeClasspathEntry.class, IJavaProject.class, boolean.class);
+            return (IRuntimeClasspathEntry[]) method.invoke(null, projectRuntimeEntry, dependencyProject, excludeTestCode);
+        } catch (Exception e) {
+            throw new GradlePluginsRuntimeException("JavaRuntime.resolveRuntimeClasspathEntry() should not be called when Buildship is installed for Eclipse 4.8", e);
+        }
+    }
+
+    private boolean hasTestAttribute(IClasspathEntry entry) {
+        for (IClasspathAttribute a : entry.getExtraAttributes()) {
+            if ("test".equals(a.getName()) && "true".equals(a.getValue())) {
```

Comment:
WDYT  `Boolean.valueOf(a.getValue())`?  That works ignoring case

Reply:
Changed.